pulsar-functions/worker/src/main/java/org/apache/pulsar/functions/worker/SchedulerManager.java

```diff
@@ -415,6 +633,34 @@ private void compactAssignmentTopic() {
         }
     }
 
+    protected int updateWorkerDrainMap() {
+        long startTime = System.nanoTime();
+        int numRemovedWorkerIds = 0;
+
+        if (drainOpStatusMap.size() > 0) {
```

Comment:
I know drainOpStatusMap is a concurrent map but we are reading and writing from different different threads.  This makes me nervous.

Reply:
I didn't quite understand the question: "Is there a reason we should synchronize this with drain operation?"

updateWorkerDrainMap() does a periodic cleanup of stale information about drained workers. We need to do the operation some time after the drain has finished, when the drained worker is removed from the cluster. Since there is currently no hook (that I know of) into the SchedulerManager when a worker is added to, or removed from the cluster, the cleanup is done through a periodic poll (updateWorkerDrainMap). 

The drain operation adds a record into the concurrent map [drainOpStatusMap] when a worker is drained. An implicit assumption in the system is that the drained worker will be removed from the system, soon, by an external orchestrator. When the drained worker is seen to be removed from the system, the drainOpStatusMap is cleaned up of stale information in the updateWorkerDrainMap() code.

Reply:
@kaushik-develop there are two independent threads read and updating drainOpStatusMap.  One is a schedule periodic task:

https://github.com/apache/pulsar/pull/12178/files#diff-343b3460561c5e794ce7351d663880e37784d37e8c0a877c9e4845b5209a8c84R568

The other is when a drain operation is triggered.  Two independent actors can be read and modifying the map concurrently.

Reply:
That is by design, and should be ok.